<html><body><p>This is the fourth status update about our work on the <a class="reference external" href="https://foss.heptapod.net/pypy/pypy/-/tree/branch/py3k">py3k branch</a>, which we<br>
can work on thanks to all of the people who <a class="reference external" href="http://morepypy.blogspot.com/2012/01/py3k-and-numpy-first-stage-thanks-to.html">donated</a> to the <a class="reference external" href="http://pypy.org/py3donate.html">py3k proposal</a>.</p><p>For various reasons, less work than usual has been done since the last status<br>
update. However, some interesting things happened anyway.</p><p>As readers know, so far we spent most of the effort in fixing all PyPy's own<br>
tests which started to fail for various py2/py3 differences.  Most of them<br>
failed for shallow reasons, e.g. syntactic changes or the int/long<br>
unifications. Others failed for subtle differences and needed a bit more care,<br>
for example the fact that unbound methods are gone in Py3k.</p><p>The good news is that finally we are seeing the light at the end of the<br>
tunnel. Most of them have been fixed. For sine other tests, we introduced the<br>
concept of "py3k-skipping": some optimizations and modules are indeed failing,<br>
but right now we are concentrating on completing the core language and so we<br>
are not interested in those.  When the core language will be done, we will be<br>
able to easily find and work on the py3k-skipped tests.  In particular, for<br>
now we disabled the <tt class="docutils literal">Int</tt> and <tt class="docutils literal">String</tt> dict strategies, which are broken<br>
because of the usual int/long unification and str vs bytes.  As for modules,<br>
for now <tt class="docutils literal">_continuation</tt> (needed for stackless) and <tt class="docutils literal">_multiprocessing</tt> do<br>
not work yet.</p><p>Another non-trivial feature we implemented is the proper cleaning of exception<br>
variables when we exit <tt class="docutils literal">except</tt> blocks.  This is a feature which touches<br>
lots of levels of PyPy, starting from <tt class="docutils literal">astcompiler</tt>, down to the bytecode<br>
interpreter. It tooks two days of headache, but at the end we made it :-).</p><p>Additionally, Amaury did a lot of improvements to <tt class="docutils literal">cpyext</tt>, which had been<br>
broken since forever on this branch.</p><p>As for the next plans, now that things are starting to work and PyPy's own<br>
tests mostly pass, we can finally start to run the compiled PyPy against<br>
CPython's test suite.  It is very likely that we will have tons of failures at<br>
the beginning, but once we start to fix them one by one, a Py3k-compatible<br>
PyPy will be closer and closer.</p></body></html>